cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -219,17 +228,27 @@ public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
 
         @Override
         public AnnotationVisitor visitParameterAnnotation(int parameter, String desc, boolean visible) {
-          if (!(authEnforceAnnotationNode != null && visible &&
-            Name.class.getName().equals(Type.getType(desc).getClassName()))) {
+          if (!(authEnforceAnnotationNode != null && visible && (SUPPORTED_PARAMETER_ANNOTATIONS.contains(desc)))) {
             return null;
           }
           // Since AuthEnforce annotation was used on this method then look for parameters with named annotation and
           // store their position.
-          // TODO: Should also pick up names from QueryParam and PathParam annotations here as needed later and also
           // decide on a precedence order.
           AnnotationNode annotationNode = new AnnotationNode(desc);
-          // store the parameter position and its annotation detail
-          parameterAnnotationNode.put(parameter, annotationNode);
+          // store the parameter position and its type with annotation detail
+          ParameterDetail parameterDetail = new ParameterDetail(parameterTypes[parameter], annotationNode);
+          if (parameterDetails.containsKey(parameter)) {
```

Comment:
This if would be slightly easier to read by using the `putIfAbsent` method in `Map` (available in Java 8).

```java
          if (parameterDetails.putIfAbsent(parameter, parameterDetail) != null) {
            // we already stored some annotation for this parameter update it if the current one is Name annotation
            // as its preferred
            if (Name.class.getName().equals(Type.getType(desc).getClassName())) {
              LOG.debug("Updating parameter details for parameter at position '{}' for method '{}' in class '{}' " +
                          "from {}' to '{}'. Since %s is preferred annotation.", parameter, methodName, className,
                        parameterDetails.get(parameter), parameterDetail);
              parameterDetails.put(parameter, parameterDetail);
            }
          }
```